presto-main/src/main/java/com/facebook/presto/execution/NodeTaskMap.java

```diff
@@ -108,87 +129,155 @@ private void addTask(RemoteTask task)
+            public TaskStatsTracker(String stat, TaskId taskId, AtomicLong totalValue)
             {
+                this.stat = requireNonNull(stat, "stat is null");
                 this.taskId = requireNonNull(taskId, "taskId is null");
+                this.totalValue = requireNonNull(totalValue, "totalValue is null");
             }
 
-            public synchronized void setPartitionedSplitCount(int partitionedSplitCount)
+            public synchronized void setValue(long value)
             {
-                if (partitionedSplitCount < 0) {
-                    int oldValue = localPartitionedSplitCount.getAndSet(0);
-                    nodeTotalPartitionedSplitCount.addAndGet(-oldValue);
-                    throw new IllegalArgumentException("partitionedSplitCount is negative");
+                if (value < 0) {
+                    long oldValue = this.value.getAndSet(0L);
+                    totalValue.addAndGet(-oldValue);
+                    throw new IllegalArgumentException(stat + " is negative");
                 }
 
-                int oldValue = localPartitionedSplitCount.getAndSet(partitionedSplitCount);
-                nodeTotalPartitionedSplitCount.addAndGet(partitionedSplitCount - oldValue);
+                long oldValue = this.value.getAndSet(value);
+                totalValue.addAndGet(value - oldValue);
             }
 
             public void cleanup()
             {
-                int leakedSplits = localPartitionedSplitCount.getAndSet(0);
-                if (leakedSplits == 0) {
+                long leakedValues = value.getAndSet(0);
+                if (leakedValues == 0) {
                     return;
                 }
 
-                log.error("BUG! %s for %s leaked with %s partitioned splits.  Cleaning up so server can continue to function.",
+                log.error("BUG! %s for %s leaked with %s %s.  Cleaning up so server can continue to function.",
                         getClass().getName(),
                         taskId,
-                        leakedSplits);
+                        leakedValues,
+                        stat);
 
-                nodeTotalPartitionedSplitCount.addAndGet(-leakedSplits);
+                totalValue.addAndGet(-leakedValues);
             }
 
             @Override
             public String toString()
             {
                 return toStringHelper(this)
                         .add("taskId", taskId)
-                        .add("splits", localPartitionedSplitCount)
+                        .add(stat, value)
                         .toString();
             }
         }
+
+        // tracks stats which are passed as accumulated (cpu time) by calculating delta / duration.
+        @ThreadSafe
+        private class AccumulatedTaskStatsTracker
+                extends TaskStatsTracker
+        {
+            private final String stat;
+            private final long windowSizeInMillis;
+            SortedMap<Long, Long> values = new TreeMap<>();
```

Comment:
Where do we clear this map?  Or is it expected that it continuously grows during the lifetime of the task?

Reply:
I wonder if the keys will come into this class in time-order (and hence already sorted).  In which case we could use a LinkedHashMap instead.

Reply:
I added the following to avoid memory leak. There is a chance that It can be out of order since updateTaskStats in HttpRemoteTask being called from several places. 

`values.headMap(values.lastKey() - windowSizeInMillis).clear();`

Reply:
It's still missing the visibility modifier.

Reply:
Also, do you want to use a fastutil `Long2LongSortedMap` to avoid boxing?